lib/ansible/modules/windows/win_data_deduplication.ps1

```diff
@@ -0,0 +1,165 @@
+$enabled = Get-AnsibleParam -obj $params -name 'enabled' -type 'bool' -default $true
+$settings = Get-AnsibleParam -obj $params -name 'settings' -type 'list' -default $null
+$dedup_job = Get-AnsibleParam -obj $params -name 'dedup_job' -type 'list' -default $null
+
+$result = @{
+	changed = $false
+}
+
+# Enable-DedupVolume -Volume <String>
+
+function Set-DataDeduplication($volume) {
+
+	try {
+		$dedup_info = Get-DedupVolume -Volume "$($volume.DriveLetter):"
+	} catch {
+		$dedup_info = $null
+	}
+
+	if($null -ne $dedup_info) {
+		$is_enabled = $dedup_info.Enabled
+	} else {
+		$is_enabled = $null
+	}
+
+	if ($is_enabled -ne $enabled) {
+		if($enabled) {
+			try {
+				Enable-DedupVolume -Volume "$($volume.DriveLetter):"
+			} catch {
+				Fail-Json $result $_.Exception.Message
+			}
+		} else {
+			try {
+				Disable-DedupVolume -Volume "$($volume.DriveLetter):"
+			} catch {
+				Fail-Json $result $_.Exception.Message
+			}
+		}
+		$result.msg += " Deduplication on volume: "+$volume.DriveLetter+" set to $enabled"
+		$result.changed = $true
+	}
+
+}
+
+
+# Set-DedupVolume -Volume <String>`
+#                 -NoCompress <bool> `
+#                 -MinimumFileAgeDays <UInt32> `
+#                 -MinimumFileSize <UInt32> (minimum 32768)
+
+function Set-DataDedupJobSettings ($volume) {
+
+	try {
+		$dedup_info = Get-DedupVolume -Volume "$($volume.DriveLetter):"
+	} catch {
+		$dedup_info = $null
+	}
+
+	ForEach ($key in $settings.keys) {
+		$update_key = $key
+		$update_value = $settings.$($key)
+
+		if ($update_key -eq "MinimumFileSize" -and $update_value -lt 32768) {
+			$update_value = 32768
+		}
+
+		try {
+			$current_value = ($dedup_info | Select-Object -ExpandProperty $update_key)
+		} catch {
+			$current_value = $null
+		}
+
+		if ($update_value -ne $current_value) {
+			$command_param = @{
+				$($update_key) = $update_value
+			}
+
+			try {
+				Set-DedupVolume -Volume "$($volume.DriveLetter):" @command_param
+				$result.msg += " Setting DedupVolume settings for $update_key"
+
+			} catch {
+				Fail-Json $result $_.Exception.Message
+			}
+
+			$result.changed = $true
+		}
+	}
+
+}
+
+# Start-DedupJob -Volume <String> `
+# 	       -Type <String>
+function Start-DataDedupJob($volume) {
+
+	$dedup_job_queue = Get-DedupJob -Volume "$($volume.DriveLetter):" -ErrorAction SilentlyContinue
+
+	if(( $dedup_job_queue | Select-Object -ExpandProperty State) -ne "Queued" ) {
+
+		$command_param = @{}
+
+		ForEach ($key in $dedup_job.keys) {
+
+			$update_key = $key
+			$update_value = $dedup_job.$($key)
+
+			$command_param.Add($($update_key), $update_value)
+
+		}
+
+		$dedup_job_start = Start-DedupJob -Volume "$($volume.DriveLetter):" @command_param
+		$result.msg += " DedupJob "+($dedup_job_start | Select-Object -ExpandProperty State)
+
+		$result.changed = $true
+	} else {
+		$result.msg += " A DedupJob is already queued for that volume."
+	}
+
+}
+
+function DataDeduplication($volume) {
```

Comment:
I really think you should move all this logic to `Set-DataDeduplication` with parameters for all of those

Reply:
Ok, I'll work on that